src/cpp/session/modules/data/SessionData.cpp

```diff
@@ -25,22 +27,168 @@ using namespace rstudio::core ;
+       continuation_(continuation),
+       request_(request)
+   {
+   }
+
+   FilePath pathFromModulesSource(std::string sourceFile)
+   {
+      FilePath modulesPath = session::options().modulesRSourcePath();
+      FilePath srcPath = modulesPath.complete(sourceFile);
+
+      return srcPath;
+   }
+
+   FilePath pathFromSource(std::string sourceFile)
+   {
+      FilePath sourcesPath = session::options().coreRSourcePath();
+      FilePath srcPath = sourcesPath.complete(sourceFile);
+
+      return srcPath;
+   }
+
+   void start()
+   {
+      std::ostringstream jsonStream;
+      json::Value requestValue = request_.params;
+      json::write(requestValue, jsonStream);
+      std::string jsonData = jsonStream.str();
+
+      std::string cmd = ".rs.preview_data_import_async('" + jsonData + "')";
+
+      std::vector<core::FilePath> sources;
+      sources.push_back(pathFromSource("Tools.R"));
+      sources.push_back(pathFromModulesSource("ModuleTools.R"));
+      sources.push_back(pathFromModulesSource("SessionDataViewer.R"));
+      sources.push_back(pathFromModulesSource("SessionDataImportV2.R"));
+
+      async_r::AsyncRProcess::start(cmd.c_str(), FilePath(), async_r::R_PROCESS_VANILLA, sources);
+   }
+
+   void onCompleted(int exitStatus)
+   {
+      if (terminationRequested())
+      {
+          // If the user requested to terminate this request, there is no need to
+          // trigger an error event.
+          return;
+      }
+
+      if (errors_.size() > 0 || exitStatus != EXIT_SUCCESS)
+      {
+         json::Object jsonError;
+
+         jsonError["message"] = "The operation failed to complete.";
+         if (errors_.size() > 0)
+         {
+            jsonError["message"] = json::toJsonArray(errors_);
+         }
+
+         json::Object jsonErrorResponse;
+         jsonErrorResponse["error"] = jsonError;
+
+         json::JsonRpcResponse response;
+         response.setResult(jsonErrorResponse);
+
+         continuation_(Success(), &response);
+
+         return;
+      }
+
+      json::Value jsonResponse;
+      json::parse(output_, &jsonResponse);
+
+      json::JsonRpcResponse response;
+      response.setResult(jsonResponse);
+
+      continuation_(Success(), &response);
+   }
+
+   void onStdout(const std::string& output)
+   {
+      output_ += output;
+   }
+
+   void onStderr(const std::string& output)
+   {
+      errors_.push_back(output);
+   }
+
+   const json::JsonRpcFunctionContinuation continuation_;
+   json::JsonRpcRequest request_;
+   std::vector<std::string> errors_;
+   std::string output_;
+};
+
+boost::shared_ptr<AsyncDataPreviewRProcess> s_pActiveDataPreview;
+
+void getPreviewDataImportAsync(
+        const json::JsonRpcRequest& request,
+        const json::JsonRpcFunctionContinuation& continuation)
+{
+   if (s_pActiveDataPreview &&
+       s_pActiveDataPreview->isRunning())
+   {
+   }
+   else
+   {
+      s_pActiveDataPreview = AsyncDataPreviewRProcess::create(request, continuation);
```

Comment:
I think it might be better to have this method return a Bool indicating whether there was already a preview running. You may not be using this now but it might be useful later on the client to know that this has occurred.